driver-core/src/main/com/mongodb/internal/connection/LoadBalancedClusterableServerFactory.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.MongoCompressor;
+import com.mongodb.MongoCredential;
+import com.mongodb.MongoDriverInformation;
+import com.mongodb.ServerAddress;
+import com.mongodb.ServerApi;
+import com.mongodb.connection.ClusterConnectionMode;
+import com.mongodb.connection.ClusterId;
+import com.mongodb.connection.ConnectionPoolSettings;
+import com.mongodb.connection.ServerId;
+import com.mongodb.connection.ServerSettings;
+import com.mongodb.connection.StreamFactory;
+import com.mongodb.event.CommandListener;
+import com.mongodb.event.ServerListener;
+
+import java.util.List;
+
+public class LoadBalancedClusterableServerFactory implements ClusterableServerFactory {
```

Comment:
We do need to document thread-safety regardless of whether a class is internal or not. This can de done either by writing a javadoc, or by using the annotation. What are the downsides of using this annotation on internal classes?

Reply:
Yes, thread-safety/immutability is not documented for all classes, and it causes problems when reading the code. So on one hand we have "preserve consistency by continuing not documenting thread-safety/immutability and, therefore, making the situation worse", and on the other hand we have "ignore consistency because it is not helpful in this case, and start making thing better by documenting thread-safety/immutability". The latter seems beneficial, while the former seems detrimental.

Reply:
Thinking about this more, I think the issue is that it's obvious to me which internal classes need to be thread safe, but it's not obvious to you.  So I'll start adding them

Reply:
Done